### PR TITLE
fix(AWS API Gateway): Deprecate invalid `apiGateway` settings

### PR DESCRIPTION
**Pull request by jkklapp**
Closes: #9139 #8921

### CODE REVIEW

**Review by pgrzesik — CHANGES_REQUESTED** [review]
Thank you @jkklapp - that's a good start - please see my comments and let me know what do you think :bow:

File: lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.js [hunk 1]
```diff
@@ -55,7 +55,23 @@ module.exports = {
         .call(this)
         .then(resolveRestApiId.bind(this))
         .then(() => {
-          if (this.apiGatewayRestApiId) return resolveDeploymentId.call(this);
+          if (this.apiGatewayRestApiId) {
```
[the remark this answers is not in the document]
Comment (jkklapp) [reply lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.js:58]: @pgrzesik just to double check on the last line

> It should check for existence of provider.apiGateway.restApiId property in conjunction with the other two settings.

Is this deprecation to be logged when **both** tracing and logs are configured or is it one or the other too?

File: docs/deprecations.md [hunk 2]
```diff
@@ -363,3 +363,11 @@ Please use `onUnauthenticatedRequest` instead. `allowUnauthenticated` will be re
...
 
 Please use `bin/serverless.js` instead. `bin/serverless` will be removed with v2.0.0
+
+<a name="NON_APPLICABLE_SETTINGS_ARE_CONFIGURED"><div>&nbsp;</div></a>
+
+## Non-applicable settings configured
+
+Deprecation code: `NON_APPLICABLE_SETTINGS_ARE_CONFIGURED`
+
+When `provider.tracing.apiGateway` or `provider.logs.restApi` are set along with `provider.apiGateway.restApiId`, these are ignored when the API Gateway is created without a stack.
```
[the remark this answers is not in the document]
Comment (jkklapp) [reply docs/deprecations.md:373]: How about

```md
When `provider.apiGateway.restApiId` is set along `provider.tracing.apiGateway` or `provider.logs.restApi` these options are ignored as the API Gateway service has been created outside of serverless.
```
?

File: docs/providers/aws/events/apigateway.md [hunk 3]
```diff
@@ -1741,7 +1741,7 @@ provider:
     restApi: true
 ```
 
-The log streams will be generated in a dedicated log group which follows the naming schema `/aws/api-gateway/{service}-{stage}`.
+The log streams will be generated in a dedicated log group which follows the naming schema `/aws/api-gateway/{service}-{stage}`. Note the property `restApi` can only be true if the API Gateway service has been created inside the stack.
```
Comment (pgrzesik) [inline docs/providers/aws/events/apigateway.md:1744]: Let's go with more descriptive documentation as the current proposal is vague. Let's start on a new line and go with something like:

```
**Note:** If external API Gateway resource is used and imported via `provider.apiGateway.restApiId` setting, `provider.logs.restApi` setting will be ignored.
```

File: docs/deprecations.md [hunk 4]
```diff
@@ -363,3 +363,11 @@ Please use `onUnauthenticatedRequest` instead. `allowUnauthenticated` will be re
 Deprecation code: `BIN_SERVERLESS`
 
 Please use `bin/serverless.js` instead. `bin/serverless` will be removed with v2.0.0
+
+<a name="AWS_API_GATEWAY_NON_APPLICABLE_SETTINGS"><div>&nbsp;</div></a>
+
+## Non-applicable settings configured for API Gateway
+
+Deprecation code: `AWS_API_GATEWAY_NON_APPLICABLE_SETTINGS`
+
```
Comment (pgrzesik) [inline docs/deprecations.md:372]: Let's go with something like:

```
When external API Gateway resource is used an imported via `provider.apiGateway.restApiId` setting, both `provider.logs.restApi` and `provider.tracing.apiGateway` will be ignored. These settings are applicable only if API Gateway resource is provisioned by Serverless Framework.
```

File: lib/plugins/aws/package/compile/events/apiGateway/index.js [hunk 5]
```diff
@@ -263,6 +263,33 @@ class AwsCompileApigEvents {
...
+        if (
+          this.serverless.service.provider.name === 'aws' &&
+          this.serverless.service.provider.apiGateway &&
+          this.serverless.service.provider.apiGateway.restApiId &&
+          this.serverless.service.provider.tracing &&
+          this.serverless.service.provider.tracing.apiGateway != null
+        ) {
+          this.serverless._logDeprecation(
+            'AWS_API_GATEWAY_NON_APPLICABLE_SETTINGS',
+            'Property "provider.tracing.apiGateway" is configured and is' +
```
Comment (pgrzesik) [inline lib/plugins/aws/package/compile/events/apiGateway/index.js:276]: Let's be more precise with this message and go with something like:

```
When external API Gateway resource is imported via `provider.apiGateway.restApiId`, property "provider.tracing.apiGateway" will be ignored.
```

File: lib/plugins/aws/package/compile/events/apiGateway/index.js [hunk 6]
```diff
@@ -263,6 +263,33 @@ class AwsCompileApigEvents {
...
+        if (
+          this.serverless.service.provider.name === 'aws' &&
+          this.serverless.service.provider.apiGateway &&
+          this.serverless.service.provider.apiGateway.restApiId &&
+          this.serverless.service.provider.logs &&
+          this.serverless.service.provider.logs.restApi != null
+        ) {
+          this.serverless._logDeprecation(
+            'AWS_API_GATEWAY_NON_APPLICABLE_SETTINGS',
+            'Property "provider.logs.restApi" is configured and is not ' +
```
Comment (pgrzesik) [inline lib/plugins/aws/package/compile/events/apiGateway/index.js:289]: Let's go with similar wording as the one proposed for `tracing` deprecation

**Review by pgrzesik — CHANGES_REQUESTED** [review]
Thanks @jkklapp - I have a few minor comments, please let me know what do you think :bow:

File: test/unit/lib/plugins/aws/package/compile/events/apiGateway/index.test.js [hunk 7]
```diff
@@ -31,6 +31,7 @@ describe('AwsCompileApigEvents', () => {
     const options = {
       stage: 'dev',
       region: 'us-east-1',
+      tracing: { apiGateway: true },
```
[the remark this answers is not in the document]
Comment (jkklapp) [reply test/unit/lib/plugins/aws/package/compile/events/apiGateway/index.test.js:34]: I was trying to make tests cover the new code, I'll revert the change

File: lib/plugins/aws/package/compile/events/apiGateway/index.js (hunk 5 above)
Comment (pgrzesik) [reply lib/plugins/aws/package/compile/events/apiGateway/index.js:276]: Looks like this wasn't addressed, am I correct?

File: lib/plugins/aws/package/compile/events/apiGateway/index.js (hunk 6 above)
Comment (pgrzesik) [reply lib/plugins/aws/package/compile/events/apiGateway/index.js:289]: Same here

File: docs/providers/aws/events/apigateway.md [hunk 8]
```diff
@@ -1741,7 +1743,7 @@ provider:
     restApi: true
 ```
 
-The log streams will be generated in a dedicated log group which follows the naming schema `/aws/api-gateway/{service}-{stage}`.
+The log streams will be generated in a dedicated log group which follows the naming schema `/aws/api-gateway/{service}-{stage}`. **Note:** If external API Gateway resource is used and imported via `provider.apiGateway.restApiId` setting, `provider.logs.restApi` setting will be ignored.
```
Comment (pgrzesik) [inline docs/providers/aws/events/apigateway.md:1746]: Looks good - let's put **Note:** on a separate line

**Review by pgrzesik — CHANGES_REQUESTED** [review]
Thank you @jkklapp - I believe a few comments from previous review weren't addressed - could you please check them out again and let me know if I missed something?

File: docs/deprecations.md [hunk 9]
```diff
@@ -363,3 +363,11 @@ Please use `onUnauthenticatedRequest` instead. `allowUnauthenticated` will be re
 Deprecation code: `BIN_SERVERLESS`
 
 Please use `bin/serverless.js` instead. `bin/serverless` will be removed with v2.0.0
+
+<a name="AWS_API_GATEWAY_NON_APPLICABLE_SETTINGS"><div>&nbsp;</div></a>
```
Comment (pgrzesik) [inline docs/deprecations.md:367]: Sorry for not noticing it before - it should be listed as first on the list, right after `How to disable specific deprecation logs` section

**Review by pgrzesik — CHANGES_REQUESTED** [review]
Thank you @jkklapp - looks good overall, I missed one minor thing and after that we should be good to go :100: 

Note: After pushing new changes, please request a review, otherwise it's hard to know if the changes are ready for a review.

**Comment by jkklapp** [comment]
@pgrzesik I moved the message in `deprecations.md` please confirm this is ok.

**Review by pgrzesik — APPROVED** [review]
Thank you @jkklapp :bow: Looks great :+1:

**Pull request merged**